src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,121 @@
+            array[size] = value;
+            size++;
+        } else {
+            array = grow(array);
+            array[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < size && index >= 0) {
+            if (size == array.length) {
+                array = grow(array);
+            }
+            System.arraycopy(array, index, array, index + 1, array.length - index - 1);
+            array[index] = value;
+            size++;
+        } else if (index == size) {
+            this.add(value);
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + size);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (list.size() + size > array.length) {
+            array = grow(array);
+        }
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + size);
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                array[i] = value;
+                return;
+            }
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index " + index
+                + " out of bounds for length " + size);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = null;
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                value = array[i];
+                array[i] = null;
+                for (int j = i + 1; j < size; j++) {
+                    array[j - 1] = array[j];
+                }
+                array[size - 1] = null;
+                size--;
+                return value;
+
+            }
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index " + index
+                + " out of bounds for length " + size);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int removedIndex = -1;
+        T value = null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(array[i], element)) {
+                value = array[i];
+                this.remove(i);
+                return value;
+            }
+        }
+        throw new NoSuchElementException("There is no such element in the list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T[] grow(T[] originalArray) {
```

Comment:
```suggestion
    private void resize() {
        if (elements.length == size) {
            int newCapacity = (int) (elements.length * CAPACITY_INDEX);
            Object[] newArray = new Object[newCapacity];
            System.arraycopy(elements, 0, newArray, 0, size);
            elements = newArray;
        }
    }
```

Where private static final double CAPACITY_INDEX = 1.5;